chown+chgrp+chmod: Fix handling of preserve root flag and error messages

BenWiederhake opened the pull request:
This makes the GNU `preserve-root.log` test green.

In particular:
- We used to have the wrong logic to detect whether or not to issue a warning. The root (punt not intended) cause was that `Path::is_dir()` completely ignores whether a given argument has a trailing slash or not, but that is *the* criterion that GNU (and many other tools) apparently uses to decide whether the path points "at" or "into" a directory. (See below for demonstration.)
- chown and chgrp also didn't correctly output an indication why the warning was being given
- And finally, `chmod` simply forgot to output its own name in one place.

Additional notes:
- This makes the uucore function `resolve_relative_path` superfluous. This function is by necessity ill-defined: Depending on the context, `..` is either the logical parent directory, sometimes the physical parent directory. This function can only work for the latter case, in which case `Path::canonicalize` is often a better approach.
- While rewriting, I noticed and documented an unavoidable TOCTOU bug. (If root runs `chown -R -L` and the attacker can quickly inject a directory and replace it with a symlink to `/`, then `--preserve-root` does not trigger. However, in this scenario the attacker could probably just as well create a handful of symlinks, one for each directory in `/`, for the same effect. So this is not a vulnerability, just an unavoidable bug.)

<details><summary><code>Path::is_dir</code> demo</summary>

```rust
fn main() {
    for name in "/tmp /tmp/ /xxx /xxx/".split(" ") {
        let buf: &std::path::Path = name.as_ref();
        println!("{:?}.is_dir() = {}", buf.as_os_str(), buf.is_dir());
    }
}
```

produces

```
"/tmp".is_dir() = true
"/tmp/".is_dir() = true
"/xxx".is_dir() = false
"/xxx/".is_dir() = false
```

BenWiederhake commented:
Changes since last push:
- Uglified the code away from `OsStr::ends_with()`, because that is Rust 1.74.0, and MSRV is Rust 1.70.0.
- I don't understand why mv/backup-dir is doing better now. I touched none of these tools O.o

tertsdiepraam commented on src/uu/chmod/src/chmod.rs:
```diff
@@ -267,7 +267,7 @@ impl Chmoder {
                     1,
                     format!(
-                        "it is dangerous to operate recursively on {}\nuse --no-preserve-root to override this failsafe",
+                        "it is dangerous to operate recursively on {}\nchmod: use --no-preserve-root to override this failsafe",
```
I wonder if this could also be fixed in general for printed errors, but that's out of scope for this PR.

tertsdiepraam commented on src/uucore/src/lib/features/perms.rs:
```diff
@@ -188,6 +189,77 @@ pub struct ChownExecutor {
+}
+
+fn is_root(path: &Path, would_recurse_symlink: bool) -> bool {
+    // Prohibit only if:
+    // (--preserve-root and -R present) &&
+    // (path canonicalizes to "/") &&
+    // (
+    //     (path is a symlink && would recurse on this symlink) ||
+    //     (path is not a symlink)
+    // )
+    // The first clause is checked by the caller, the second and third clause is checked here.
+    // The caller has to evaluate -P/-H/-L into 'would_recurse_symlink'.
+    // Recall that canonicalization resolves both relative paths (e.g. "..") and symlinks.
```
I think this makes sense as a docstring. We don't really need to mention `--preserve-root` and `-R` here, because we're not in the context of a specific util.

BenWiederhake replied:
Why not mention `--preserve-root` in the function that has the express purpose of the flag? Also, this serves as a reminder of how to use this function correctly.

I'm happy to turn this a docstring.

tertsdiepraam replied:
I guess mentioning them is fine but it should then also mention on which utils that flag is used.

BenWiederhake replied:
Good idea, will do!

tertsdiepraam commented on src/uucore/src/lib/features/perms.rs:
```diff
@@ -586,3 +645,73 @@ pub fn chown_base(
+
+#[cfg(test)]
+mod tests {
+    // Note this useful idiom: importing names from outer (for mod tests) scope.
```
Yeah, we know :)

BenWiederhake replied:
Sorry, I copy-pasted it from somewhere else. I'll delete it :)

tertsdiepraam commented on src/uucore/src/lib/features/perms.rs:
```diff
@@ -188,6 +189,77 @@ pub struct ChownExecutor {
+    if let Ok(p) = path.canonicalize() {
+        let path_buf = path.to_path_buf();
+        if p.parent().is_none() {
+            match path_buf.to_str() {
```
We can print osstrs with quotes using the Quotable trait: https://docs.rs/uucore/latest/uucore/display/trait.Quotable.html

BenWiederhake replied:
The GNU tests check for specific error messages, and this match statement just accesses the internal `&OsStr` to check for a specific value. What do you mean, which improvement did I miss?

tertsdiepraam replied:
Yes, the messsage is correct, but I think the conversion to a `str` can be avoided with that quotable trait. It can print `OsStr` directly. But maybe there's something I'm missing?

BenWiederhake replied:
Ah! I think I see what you mean. Yes, I'll address it.

tertsdiepraam commented on src/uucore/src/lib/features/perms.rs:
```diff
@@ -188,6 +189,77 @@ pub struct ChownExecutor {
+        }
+    }
+
+    // FIXME: TOCTOU bug! canonicalize() runs at a different time than WalkDir's recursion decision.
```
This is actually really common in these utils, because it's somewhat inherent to filesystem APIs, but indeed probably good to note.

tertsdiepraam commented on src/uucore/src/lib/features/perms.rs:
```diff
@@ -188,6 +189,77 @@ pub struct ChownExecutor {
+    is_root(path, would_recurse_symlink)
+}
+
+fn is_root(path: &Path, would_recurse_symlink: bool) -> bool {
```
Shouldn't it be `would_traverse_symlink` or is there some actual recursion going on here?

BenWiederhake replied:
You're right! I'll change the name of the variable.

tertsdiepraam commented on src/uucore/src/lib/features/perms.rs:
```diff
@@ -188,6 +189,77 @@ pub struct ChownExecutor {
+            Some(".") | Some("..") => true,
+            Some(path_str) => {
+                (path_str.ends_with(MAIN_SEPARATOR_STR))
+                    || (path_str.ends_with(&format!("{}.", MAIN_SEPARATOR_STR)))
+                    || (path_str.ends_with(&format!("{}..", MAIN_SEPARATOR_STR)))
```
How would this work when the separator is escaped for instance?

BenWiederhake replied:
I'm not sure what you mean by "escaped", since this should be the already-parsed path. If we're on unix and the string contains some backslash for example, it is an invalid path, because backslashes cannot contain backslashes on unix. If we're on Windows, the semantics of `--preserve-root` go out of the window anyway (pun not intended). Did I miss a potentially-problematic case?

tertsdiepraam replied:
I guess not, whenever we do string manipulation like this it usually fails on some escape sequence, but you might be right that there are none here.

BenWiederhake replied:
Ah, I see what you mean. Yes, I'm also uncomfortable with doing this amount of "string-based file system logic". I *think* I didn't miss anything. On the other hand, this is a safety mechanism to prevent bone-headed command-line typos, not a security measure that has to be absolutely waterproof, so a "small" bug in this decision logic shouldn't be a catastrophe.

tertsdiepraam reviewed (commented):
Alright, please check that I understand correctly. Here's what I got from this:

We need to know whether symlinks will be dereferenced on the path to know how to check for root. Dereferencing happens if either the setting for dereferencing is passed or if the `/` is at the end, so that we need to enter the directory and to do that need to dereference.

Does that make sense? If so, could you add that last part in a comment somewhere. It took me a while to understand :)

I also wonder if these path checks are the right way to do it. Maybe there's some other way to check this? It all feels inherently finnicky.

BenWiederhake commented:
Changes since last push:
- rebased on current `main`
- Converted the big introduction in `fn is_root` to a docstring
- In it, mentioned that this is specific to chown and chgrp
- Simplified the to_str/match/print handling in case the check triggers
- Renamed the argument to would_traverse_symlink. All as suggested by @tertsdiepraam

BenWiederhake commented:
Nope, wait, I got something wrong with the quoting.

BenWiederhake commented:
Changes since last push:
- Fixed the double-quoting issue (I used `"blah '{}' blah", pathbuf.quote()`, which results in the quotes appearing twice. Doink!)

sylvestre commented:
well done :)
